[Mission2. JPA] 2단계 - 연관 관계 매핑

gregolee opened the pull request:
안녕하세요. 강현구 리뷰어님!! 🙇  👋  😄 
연관 관계 매핑 리뷰👀  요청 드립니다! 🙇 
최근 아프셨다고 😷  강의 도중 들었는데, 몸 조심하시고 건강🏃‍♂️ 하시길 바랍니다!!

gregolee committed:
docs, chore: 2단계 markdown 추가, checkstyle & editorconfig 수정

README.md
- step02.md 링크할 수 있도록 수정

markdown/step02.md
- 요구사항 정리
- 학습 내용 정리
- TODO List 수정

config/naver-checkstyle-rules.xml
- 한글 메서드 입력에도 valid할 수 있도록 변경

.editorconfig
- 윈도우 환경에서 line separators를 CRLF -> LF로 변경

Resolve:
See Also:

gregolee committed:
chore: editorconfig 수정

gradlew.bat
- line_of_end 수정으로 인한 변경

Resolve:
See Also:

gregolee committed:
feat: 연관 관계 매핑 구현 완료

- [x] 4.구현
    - [x] 4-1.테스트 코드 작성
        - [x] 4-1-1.`AnswerTest`
        - [x] 4-1-2.`DeleteHistoryTest`
        - [x] 4-1-3.`QuestionTest`
        - [x] 4-1-4.`UserTest`
    - [x] 4-2.연관 관계 매핑
        - [x] 4-2-1.`Answer`
        - [x] 4-2-2.`DeleteHistory`
        - [x] 4-2-3.`Question`
        - [x] 4-2-4.`User`

Resolve:
See Also:

gregolee committed:
test: Build, CheckStyle 등 테스트

Gradle Build 확인

CheckStyle 미적용된 코드 확인
- 메서드 내 연속된 대문자 반복하지 않도록 변경
- checkstyle 내 메서드 문구 validation 조건 변경
- 라인당 120자 제한 준수

기능요구사항 및 프로그래밍 요구사항 충족여부 확인

Resolve:
See Also:

gregolee committed:
chore: 인수인계 준비

- markdown 확인
- 최종 테스트 확인
- 코드리뷰 진행

Resolve:
See Also:

kang-hyungu commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -29,14 +34,16 @@
+	@JoinColumn(name = "wirter_id", foreignKey = @ForeignKey(name = "fk_answer_writer"))
+	private User writer;
 
 	public Answer() {
```
기본생성자를 public으로 열어둘 필요가 있을까요?
protected로 바꿔보면 어떨까요?

gregolee replied:
전반적으로 엔티티에 public으로 열어둔 부분이 많았습니다. 😭 
생성자가 많은 것은 문제가 되지 않으나,
관리되지 않은 생성자 접근하여 생성하지 못하도록 하겠습니다! 🙇

kang-hyungu commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -77,20 +84,20 @@ public void setId(Long id) {
 
-	public void setWriterId(Long writerId) {
-		this.writerId = writerId;
+	public void setWriter(User writer) {
```
객체지향 생활체조
규칙 9: 게터/세터/프로퍼티를 쓰지 않는다.

필요한 getter를 제외하고 모든 도메인에서 getter, setter는 제거해보아요

gregolee replied:
다음 단계에서 적용하는 것이라 생각하고 적용하지 않았습니다.
해당 부분도 getter, setter를 제거하여 바른 메서드 명으로 고쳐 넣도록 하겠습니다! 😄

kang-hyungu commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -29,14 +34,16 @@
+	private Question question;
 
 	@Column(name = "updated_at")
 	private LocalDateTime updatedAt;
```
createdAt, updatedAt가 Question, User 클래스에도 반복적으로 나오고 있네요

1. BaseEntity라는 추상 클래스를 추가해서 중복된 필드를 모아보고
2. BaseEntity에 jpa auditing을 적용해보아요

gregolee replied:
데이터 관점으로 생각을 하고 객체지향적으로 생각해보지 못한 것 같습니다 😢 
해당 부분을 객체 지향적으로 생각해보도록 하겠습니다 🙇

kang-hyungu commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -3,28 +3,73 @@
 
 @DataJpaTest
-@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
 public class AnswerTest {
```
AnswerTest는 Answer 클래스에 대한 테스트만 남겨두시고 
AnswerRepositoryTest를 새로 추가하셔서 DataJpaTest 와 관련된 테스트를 진행해주세요
두 테스트가 섞이지 않아야 합니다.
AnwserTest는 순수 자바와 관련된 유닛 테스트라 스프링에 의존하지 않고
AnswerRepositoryTest는 현재 스프링을 사용하고 있으니 스프링에 의존한 테스트가 나오겠죠.

gregolee replied:
> AnswerTest는 Answer 클래스에 대한 테스트만 남겨두시고
> AnswerRepositoryTest를 새로 추가하셔서 DataJpaTest 와 관련된 테스트를 진행해주세요
> 두 테스트가 섞이지 않아야 합니다.
> AnwserTest는 순수 자바와 관련된 유닛 테스트라 스프링에 의존하지 않고
> AnswerRepositoryTest는 현재 스프링을 사용하고 있으니 스프링에 의존한 테스트가 나오겠죠.

단순히 비즈니스 로직의 구분으로 크게 한 곳에 묶어야 한다고 생각했습니다. 🤔 💭 
같은 비즈니스라 묶어서 관리하는 것이 아닌 테스트 영역에 따른 구분 중요하게 따지도록 하겠습니다! 😄

kang-hyungu commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -3,28 +3,73 @@
+
+	@DisplayName("테스트 초기화")
+	@BeforeEach
+	void setup() {
```
초기에 모든 객체를 만들고 삭제할 필요가 있을까요?
각 테스트 마다 필요한 객체를 만들도록 하시고 데이터는 각 테스트가 끝나면 롤백되서 삭제할 필요 없을거에요
DataJpaTest를 보시면 트랜잭션 어노테이션이 붙어있습니다.

gregolee replied:
> 초기에 모든 객체를 만들고 삭제할 필요가 있을까요?
> 각 테스트 마다 필요한 객체를 만들도록 하시고 데이터는 각 테스트가 끝나면 롤백되서 삭제할 필요 없을거에요

1. 제가 기존 행동을 답습한 코멘트를 주셨습니다. 제가 반성하고 고칠 부분 알려주시어 감사합니다! 🙇 

> DataJpaTest를 보시면 트랜잭션 어노테이션이 붙어있습니다.

2. 구현하기 급급하여 실제 내부 코드를 보지 않으려는 악습이 그대로 드러났습니다. 😢 
내부 코드도 확인하여 올바른 기능을 학습하려는 자세를 갖추도록 자각하겠습니다! 🔥

kang-hyungu commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -35,21 +80,20 @@ void save() {
 		//then
 		assertAll(
```
너무 모든걸 한꺼번에 테스트하려고 하진 않나요?
1. 객체의 모든 필드가 정상인지 테스트하는게 꼭 필요한가요? 어떤 의미가 있나요?
2. 만약 필요한 테스트라면 equals와 hashCode 메서드를 구현해서 객체의 동일성을 비교하도록 만들면 테스트 코드도 깔끔해지지 않을까요?

gregolee replied:
> 너무 모든걸 한꺼번에 테스트하려고 하진 않나요?
> 
> 1. 객체의 모든 필드가 정상인지 테스트하는게 꼭 필요한가요? 어떤 의미가 있나요?
> 2. 만약 필요한 테스트라면 equals와 hashCode 메서드를 구현해서 객체의 동일성을 비교하도록 만들면 테스트 코드도 깔끔해지지 않을까요?

네 리뷰어님께서 전달하신 말씀이 옳다고 생각합니다. 👍 🙇 
다만 생각이 많아지게 만드는 코멘트를 주셨습니다. 🤔 💭 
평소 얼마나 중복된 코드 제거와 깔끔한 코드를 작성하려는 노력이 부족했다는 생각이 듭니다.
해당 말씀을 반영하여 좀 더 나은 코드를 작성할 수 있도록 노력해보겠습니다! 🔥

kang-hyungu reviewed (changes requested):
재영님 2단계 구현하신 내용 확인했어요
몇 가지 피드백 남겼으니 참고하세요
수정하시면서 궁금한 점 생기면 코멘트 남기시거나 dm 주세요

gregolee commented:
> 재영님 2단계 구현하신 내용 확인했어요
> 몇 가지 피드백 남겼으니 참고하세요
> 수정하시면서 궁금한 점 생기면 코멘트 남기시거나 dm 주세요

리뷰 감사합니다! 👍 🥇 🙇 
코멘트를 하나하나 곰곰이 뜯어보며
고쳐야할 습관들에 대해 고민을 할 수 있었습니다. 🤔 💭 

이번 단계를 진행하며 문제점은 다음과 같습니다. 🗒️ 
1. 기존의 엔티티를 수정하지 않고 진행하려 했으며,
2. 코드 중복 제거에 대한 노력 필요하고
3. 테스트 영역에 대한 세분화 필요했습니다.

다음 리뷰 재요청에는 말씀하신 부분 깔끔하게 잘 반영하도록 해보겠습니다! 🔥 🔥 🔥 🔥 🔥

gregolee committed:
refactor: 생성자 접근제어자 `public` -> `protected` 변경, getter/setter 제거

- [ ] 4.구현
    - [x] 4-1.테스트 코드 작성
        - [x] 4-1-1.`AnswerTest`
        - [x] 4-1-2.`DeleteHistoryTest`
        - [x] 4-1-3.`QuestionTest`
        - [x] 4-1-4.`UserTest`
    - [x] 4-2.연관 관계 매핑
        - [x] 4-2-1.`Answer`
        - [x] 4-2-2.`DeleteHistory`
        - [x] 4-2-3.`Question`
        - [x] 4-2-4.`User`
    - [ ] 4-3.리뷰어님 피드백 반영
        - [x] 4-3-1.생성자 접근제어자 `public` -> `protected` 변경
        - [x] 4-3-2.getter/setter 제거
        - [ ] 4-3-3.반복적인 필드 추상클래스 활용
            - [ ] 4-3-3-1.중복된 필드 모아놓기
            - [ ] 4-3-3-2.jpa auditing 적용
        - [ ] 4-3-4.테스트 영역 구분
            - [ ] 4-3-4-1.`AnswerTest` -> `AnswerTest`, `AnswerRepositoryTest`
            - [ ] 4-3-4-2.`DeleteHistoryTest` -> `DeleteHistoryTest`, `DeleteHistoryRepositoryTest`
            - [ ] 4-3-4-3.`QuestionTest` -> `QuestionTest`, `QuestionRepositoryTest`
            - [ ] 4-3-4-4.`UserTest` -> `UserTest`, `UserRepositoryTest`
        - [ ] 4-3-5.테스트 초기화 수정
            - [ ] 4-3-5-1.`AnswerTest`
            - [ ] 4-3-5-2.`AnswerRepositoryTest`
            - [ ] 4-3-5-3.`DeleteHistoryTest`
            - [ ] 4-3-5-4.`DeleteHistoryRepositoryTest`
            - [ ] 4-3-5-5.`QuestionTest`
            - [ ] 4-3-5-6.`QuestionRepositoryTest`
            - [ ] 4-3-5-7.`UserTest`
            - [ ] 4-3-5-8.`UserRepositoryTest`
        - [ ] 4-3-6.불필요한 테스트 코드 정리
            - [ ] 4-3-6-1.`AnswerTest`
            - [ ] 4-3-6-2.`AnswerRepositoryTest`
            - [ ] 4-3-6-3.`DeleteHistoryTest`
            - [ ] 4-3-6-4.`DeleteHistoryRepositoryTest`
            - [ ] 4-3-6-5.`QuestionTest`
            - [ ] 4-3-6-6.`QuestionRepositoryTest`
            - [ ] 4-3-6-7.`UserTest`
            - [ ] 4-3-6-8.`UserRepositoryTest`

Resolve:
See Also:

gregolee committed:
chore: java convention 변경

java convention 변경
- naver-checkstyle 제거
- IDE 내 google java convention 적용
- 코드 전체 google java convention 적용

Resolve:
See Also:

gregolee committed:
refactor: 반복적인 필드 추상클래스 활용

- [ ] 4-3.리뷰어님 피드백 반영
        - [x] 4-3-1.생성자 접근제어자 `public` -> `protected` 변경
        - [x] 4-3-2.getter/setter 제거
        - [x] 4-3-3.반복적인 필드 추상클래스 활용
            - [x] 4-3-3-1.중복된 필드 모아놓기
            - [x] 4-3-3-2.jpa auditing 적용
        - [ ] 4-3-4.테스트 영역 구분
            - [ ] 4-3-4-1.`AnswerTest` -> `AnswerTest`, `AnswerRepositoryTest`
            - [ ] 4-3-4-2.`DeleteHistoryTest` -> `DeleteHistoryTest`, `DeleteHistoryRepositoryTest`
            - [ ] 4-3-4-3.`QuestionTest` -> `QuestionTest`, `QuestionRepositoryTest`
            - [ ] 4-3-4-4.`UserTest` -> `UserTest`, `UserRepositoryTest`
        - [ ] 4-3-5.테스트 초기화 수정
            - [ ] 4-3-5-1.`AnswerTest`
            - [ ] 4-3-5-2.`AnswerRepositoryTest`
            - [ ] 4-3-5-3.`DeleteHistoryTest`
            - [ ] 4-3-5-4.`DeleteHistoryRepositoryTest`
            - [ ] 4-3-5-5.`QuestionTest`
            - [ ] 4-3-5-6.`QuestionRepositoryTest`
            - [ ] 4-3-5-7.`UserTest`
            - [ ] 4-3-5-8.`UserRepositoryTest`
        - [ ] 4-3-6.불필요한 테스트 코드 정리
            - [ ] 4-3-6-1.`AnswerTest`
            - [ ] 4-3-6-2.`AnswerRepositoryTest`
            - [ ] 4-3-6-3.`DeleteHistoryTest`
            - [ ] 4-3-6-4.`DeleteHistoryRepositoryTest`
            - [ ] 4-3-6-5.`QuestionTest`
            - [ ] 4-3-6-6.`QuestionRepositoryTest`
            - [ ] 4-3-6-7.`UserTest`
            - [ ] 4-3-6-8.`UserRepositoryTest`

Resolve: #134
See Also:

gregolee committed:
refactor: 테스트 영역 구분, 테스트 초기화 수정, 불필요한 테스트 코드 정리

- [x] 4-3.리뷰어님 피드백 반영
        - [x] 4-3-1.생성자 접근제어자 `public` -> `protected` 변경
        - [x] 4-3-2.getter/setter 제거
        - [x] 4-3-3.반복적인 필드 추상클래스 활용
            - [x] 4-3-3-1.중복된 필드 모아놓기
            - [x] 4-3-3-2.jpa auditing 적용
        - [x] 4-3-4.테스트 영역 구분
            - [x] 4-3-4-1.`AnswerTest` -> `AnswerTest`, `AnswerRepositoryTest`
            - [x] 4-3-4-2.`DeleteHistoryTest` -> `DeleteHistoryTest`, `DeleteHistoryRepositoryTest`
            - [x] 4-3-4-3.`QuestionTest` -> `QuestionTest`, `QuestionRepositoryTest`
            - [x] 4-3-4-4.`UserTest` -> `UserTest`, `UserRepositoryTest`
        - [x] 4-3-5.테스트 초기화 수정
            - [x] 4-3-5-1.`AnswerTest`
            - [x] 4-3-5-2.`AnswerRepositoryTest`
            - [x] 4-3-5-3.`DeleteHistoryTest`
            - [x] 4-3-5-4.`DeleteHistoryRepositoryTest`
            - [x] 4-3-5-5.`QuestionTest`
            - [x] 4-3-5-6.`QuestionRepositoryTest`
            - [x] 4-3-5-7.`UserTest`
            - [x] 4-3-5-8.`UserRepositoryTest`
        - [x] 4-3-6.불필요한 테스트 코드 정리
            - [x] 4-3-6-1.`AnswerTest`
            - [x] 4-3-6-2.`AnswerRepositoryTest`
            - [x] 4-3-6-3.`DeleteHistoryTest`
            - [x] 4-3-6-4.`DeleteHistoryRepositoryTest`
            - [x] 4-3-6-5.`QuestionTest`
            - [x] 4-3-6-6.`QuestionRepositoryTest`
            - [x] 4-3-6-7.`UserTest`
            - [x] 4-3-6-8.`UserRepositoryTest`

Resolve: #134
See Also:

gregolee committed:
test: Build, CheckStyle 등 테스트

- Gradle Build 확인
- 기능요구사항 및 프로그래밍 요구사항 충족여부 확인

Resolve:
See Also:

gregolee committed:
chore: 인수인계 준비

- markdown 확인
- 최종 테스트 확인
- 코드리뷰 진행

Resolve:
See Also:

gregolee commented:
안녕하세요. 강현구 리뷰어님! 🙇 👍 
열심히 반영한다고 노력했지만, 부족한 부분이 있으리라 생각합니다. 😭 
이번 리뷰 요청 후 피드백할 사항이 있으시면 많이 알려주시기 바랍니다!!!! 😄 👍 💯

kang-hyungu commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,115 +1,71 @@
-import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
 
-@DataJpaTest
 @TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
테스트 메서드에 실행 순서와 서로 의존하게 작성하지 마세요
순수하게 삭제 기능만 테스트 하려는데 다른 테스트를 실행해야되면 빠르게 테스트 확인이 어렵지 않을까요?

gregolee replied:
> 테스트 메서드에 실행 순서와 서로 의존하게 작성하지 마세요
> 순수하게 삭제 기능만 테스트 하려는데 다른 테스트를 실행해야되면 빠르게 테스트 확인이 어렵지 않을까요?

네 알겠습니다
순서를 주고 작성했던 것은 작성된 순서대로 보여주기 위함이었습니다 😢
근데 그렇게 시작한 순서가 저도 모르게 의존하게 만들게끔 작성하게 되었습니다. 😢😢😢
다음 단계에서는 해당 의존성을 제거하여 작성토록 하겠습니다!
리뷰 감사합니다!!!

kang-hyungu commented on src/test/java/qna/domain/DeleteHistoryRepositoryTest.java:
```diff
@@ -0,0 +1,120 @@
+		//given
+
+		//when
+		DeleteHistory actual = deleteHistories.save(deleteHistory1);
```
클린코드 25p

의미 있게 구분하라
컴파일러나 인터프리터만 통과하려는 생각으로 코드를 구현하는 프로그래머는 스스로 문제를 일으킨다.
예를 들어, 동일한 범위 안에서 다른 두 개념에 같은 이름을 사용하지 못한다.
...

컴파일러를 통과할지라도 연속된 숫자를 덧붙이거나 불용어(noise word)를 추가하는 방식은 적절하지 못하다.
이름이 달라야 한다면 의미도 달라져야 한다.

연속적인 숫자를 덧붙인 이름(a1, a2 ... aN)은 의도적인 이름과 정반대다.
이런 이름은 그릇된 정보를 제공하는 이름도 아니며, 아무런 정보를 제공하지 못하는 이름일 뿐이다.
저자 의도가 전혀 드러나지 않는다.


클린코드 157p

테스트 코드는 실제 코드 못지 않게 중요하다. 테스트 코드는 이류 시민이 아니다.
테스트 코드는 사고와 설계와 주의가 필요하다.
실제 코드 못지 않게 깨끗하게 짜야한다.

테스트는 유연성, 유지보수성, 재사용성을 제공한다. 테스트 케이스가 있으면 변경이 쉬워지기 때문이다.
따라서 테스트 코드가 지저분하면 코드를 변경하는 능력이 떨어지며 코드 구조를 개선하는 능력도 떨어진다.
테스트 코드가 지저분할수록 실제 코드도 지저분해진다.
결국 테스트 코드를 잃어버리고 실제 코드도 망가진다.

gregolee replied:
지난 미션에서도 부족한 부분이었습니다. 😢 
이번 미션에서는 하지 않도록 노력했으나, 역시나 리뷰어님께는 바로 걸리고 말았네요 👍 🥇 💯 
좋은 지적 감사드리며 다음 단계에서 해당 부분 반영하겠습니다! 🔥 
감사합니다!! 🙇

kang-hyungu reviewed (approved):
재영님 피드백 반영 잘해주셨네요
pr은 머지 처리하겠습니다. 다음 단계 진행해주세요. 수고하셨습니다
